Update MASTG-KNOW-0030 and MASTG-KNOW-0032 (#3790)
OWASP/mastg: knowledge/android/MASVS-RESILIENCE/MASTG-KNOW-0032.md

Unlike @MASTG-KNOW-0030, which covers artifact-based detection (e.g., scanning for tool-specific strings or checking for open ports), this document focuses on detecting the _modifications_ that instrumentation tools make to the app's code and memory.

!!! note
  Runtime integrity verification is inherently a cat-and-mouse game. Detection methods and bypass techniques evolve continuously. Determined attackers with sufficient time and resources can typically circumvent these protections, especially on rooted devices (see [Tan, 2016](https://blackhat.com/docs/us-16/materials/us-16-Tan-Bad-For-Enterprise-Attacking-BYOD-Enterprise-Mobile-Security-Solutions-wp.pdf)). These techniques should be part of a defense-in-depth strategy, not a standalone solution.

The following sections present these techniques grouped into four categories based on the type of integrity violation they detect. Each category includes specific methods and a discussion of their effectiveness and potential bypasses. They help answer the following questions:

> | Question | Category | Techniques |

Reviewer: If I understand correctly, the use of the word "Techniques" in this KNOW does not have the same meaning as the techniques the MASTG records at https://mas.owasp.org/MASTG/techniques/. The way I see it is that the techniques in the MASTG are aimed towards performing tests whereas this use of the word "Techniques" would be closer to a best practice for the developer.

With that being said, Shouldn't each of these techniques just point to their own BEST? Just wondering, correct me if you see it differently 👍

Reviewer: Techniques -> Defensive controls

Installing such hooks requires the target memory region to be writable. The kernel records current memory permissions in [`/proc/self/maps`](https://man7.org/linux/man-pages/man5/proc_pid_maps.5.html). Under normal conditions, code sections (marked `r-xp`) are never writable, so an executable mapping that is also writable, such as `rwxp`, is suspicious. A writable mapping is also suspicious when it corresponds to a region expected to be read-only after relocation, such as RELRO-protected data. For RELRO-protected regions, the [expected state after relocation is read-only](https://android.googlesource.com/platform/bionic/%2B/master/linker/linker.cpp). Ordinary `rw-p` library data mappings are expected and should not be flagged by themselves. Checking for suspicious permissions provides a broad pre-filter before performing the specific pointer verification below.

### PLT/GOT Hook Detection

> The Procedure Linkage Table / Global Offset Table (PLT/GOT) stores relocation-resolved function addresses used by dynamically linked calls. At runtime, the dynamic linker patches these entries with the absolute addresses of imported functions. _PLT/GOT hooks_ overwrite the stored function addresses, redirecting legitimate function calls to adversary-controlled code (e.g., using libraries such as [xHook](https://github.com/iqiyi/xHook)). This type of hook can be detected by enumerating the process memory map and verifying that each relevant PLT/GOT entry points to the expected symbol implementation or to an allowed executable mapping.

Reviewer: Shouldn't we register xHook as a tool instead?

Author: Since it's just one mention we'll leave it like this for now.

!!! note

>     `ArtMethod` structure layout varies across Android versions, requiring version-specific offset handling.

Reviewer: ```suggestion
    There is no public API to obtain a reference to the `ArtMethod` structure that backs a Java method. JNI's `FromReflectedMethod` returns a `jmethodID` which could be reinterpreted as an `ArtMethod*`, but this depends on ART configuration and it is not a guarantee. On runtimes using opaque/index JNI IDs, it is not a raw pointer. The `ArtMethod` layout itself also varies across Android versions, requiring version-specific offset handling that is brittle and error-prone.
```

> When a framework hooks a method, it may replace the original entry point with a pointer to hook or bridge code. Detection approaches include:

Reviewer: ```suggestion
When a framework hooks a method, it may replace the original entry points with a pointer to hook or bridge code. Detection approaches include:
```

> - **Entry point verification**: Using JNI's `FromReflectedMethod` to obtain the `ArtMethod` pointer and verify the entry point falls within legitimate regions (OAT file, interpreter, or JIT code cache)

Reviewer: We should clarify here it is not enough getting the pointer to the `ArtMethod` structure, you still need to get the offset to the `entry_point_from_quick_compiled_code_` field to inspect if it now points to the Frida trampoline. Last time I checked, frida does this by doing pointer enumeration on the `ArtMethod` structure leveraging some known fields.

It is a bit brittle since the offsets vary between Android versions.

Reviewer: See `_getArtMethodSpec` in Frida https://github.com/frida/frida-java-bridge/blob/6ab2a63e4b02ab881f11ea06632cebea4b1b435e/lib/android.js#L1172

Reviewer: Let's stick with adding clarification that you would still need to obtain the pointer to the entry point and that since it's a private and thus, unstable API.

> - **Entry point verification**: Using JNI's `FromReflectedMethod` to obtain the `ArtMethod` pointer and verify the entry point falls within legitimate regions (OAT file, interpreter, or JIT code cache)

Reviewer: ```suggestion
- **Entry point verification**: Inspect the relevant `ArtMethod` entrypoint fields for the target Android version and verify that they fall within legitimate regions (OAT file, interpreter bridge, JNI/native stubs, or JIT code cache)
```

> - **Access flags inspection**: Check if `kAccNative` (0x0100) was unexpectedly set

Reviewer: Same with this, IIRC apps can't get stable field access to `ARTMethod` fields.

> - **Trampoline detection**: Scan the entry point for known hook signatures

Reviewer: ```suggestion
- **Trampoline detection**: Scan the entry points for known hook signatures
```
